packages/gensx/src/checkpoint.ts

```diff
@@ -0,0 +1,180 @@
+      return v.toString(16);
+    });
+  }
+}
+
+export interface ExecutionNode {
+  id: string;
+  componentName: string;
+  parentId?: string;
+  startTime: number;
+  endTime?: number;
+  props: Record<string, unknown>;
+  output?: unknown;
+  children: ExecutionNode[];
+  metadata?: {
+    logs?: string[];
+    tokenCounts?: {
+      input: number;
+      output: number;
+    };
+    [key: string]: unknown;
+  };
+}
+
+export interface CheckpointWriter {
+  currentNode?: ExecutionNode;
+  root?: ExecutionNode;
+  addNode: (node: Partial<ExecutionNode>) => Promise<string>;
+  completeNode: (id: string, output: unknown) => void;
+  addMetadata: (id: string, metadata: Record<string, unknown>) => void;
+  write: () => void;
+  checkpointsEnabled: boolean;
+}
+
+export class CheckpointManager implements CheckpointWriter {
+  private nodes = new Map<string, ExecutionNode>();
+  public root?: ExecutionNode;
+  public currentNode?: ExecutionNode;
+  public checkpointsEnabled: boolean;
+
+  // Track active checkpoint write
+  private activeCheckpoint: Promise<void> | null = null;
+  private pendingUpdate = false;
+
+  constructor() {
+    // Set checkpointsEnabled based on environment variable
+    // Environment variables are strings, so check for common truthy values
+    const checkpointsEnv = process.env.GENSX_CHECKPOINTS?.toLowerCase();
+    this.checkpointsEnabled =
+      checkpointsEnv === "true" ||
+      checkpointsEnv === "1" ||
+      checkpointsEnv === "yes";
+  }
+
+  // updateCheckpoint POSTs the current component tree to the GenSX API.
+  // Special care is taken to do this in a non-blocking manner, and in a way that
+  // minimizes chattiness with the server.
+  // We only write one checkpoint at a time. If another checkpoint comes in while
+  // we're still processing, we just mark that we need another update and return.
+  // When the current checkpoint write is complete, we check if there was a pending
+  // update request, and if so, we trigger another write.
+  private updateCheckpoint() {
+    if (!this.checkpointsEnabled || !this.root) {
+      return;
+    }
+
+    // If there's already a pending update, just mark that we need another update
+    if (this.activeCheckpoint) {
+      this.pendingUpdate = true;
+      return;
+    }
+
+    // Start a new checkpoint write
+    this.activeCheckpoint = this.writeCheckpoint().finally(() => {
+      this.activeCheckpoint = null;
+
+      // If there was a pending update requested while we were writing,
+      // trigger another write
+      if (this.pendingUpdate) {
+        this.pendingUpdate = false;
+        this.updateCheckpoint();
+      }
+    });
+  }
+
+  private async writeCheckpoint() {
+    if (!this.root) return;
+
+    try {
+      const response = await fetch("http://localhost:3000/api/execution", {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(this.root),
```

Comment:
We write the entire tree, every time?

Reply:
Yup. We can replace this with a journaling API in the future if we find performance to be an issue. However, all of these calls are non-blocking and we make the minimal number of requests necessary.